CMake: Add guards against detecting system packages

johnwparent opened the pull request:
**Work in Progress, non draft for Gitlab testing purposes**

CMake has a powerful package detection framework via `find_package`. Numerous Spack packages take advantage of this, and Spack itself of course leverages the `CMAKE_PRFEFIX_PATH` to populate that variable with packages in a given DAG so CMake can detect them. However, CMake is almost too good at finding packages available on the system, which is bad for Spack, as it allows for the subtle, hard to detect pollution of Spack build environments with non Spack controlled packages.

Consider the scenario:

1. A package has a new version added.
1. A Spack dev updates that package's version, and attempts to build
1. Unbeknownst to our Spack dev, this new version has a dependency on `zlib`
1. The package with a new version is a `CMakePackage`, the Spack dev has a system `zlib` at `/usr/lib/zlib`
1. At this point, attempting to build the new version should fail because Spack is not vendoring `zlib`, however...
1. The build succeeds because CMake find the system `zlib` and uses that to fulfill the requirement
1. The Spack dev, unaware of the new requirement for `zlib` pushes the new version and its is merged
1. A user tries to build this new version and has an incompatible version of `zlib` on their system
1. They get a bunch of weird build errors griping about zlib, but when they run `spack solve` or `spack spec` zlib is not in the DAG


Outside of that scenario (which actually happened), Spack should not be allowing system packages to infiltrate our build contexts.

TODO:
- [x] Document why each CMake variable is defined for sandboxing
- [x] Ensure `<PKG>_ROOT` and `<PKG>_DIR` env variables are sanitized
- [x] Ensure `CMAKE_FRAMEOWKR_PATH` and `CMAKE_APPBUNDLE_PATH` env vars are purged
- [x] Ensure packages that rely on these mechanisms to consume other Spack packages still can 


<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

johnwparent committed:
CMake: Add guards against detecting system packages

johnwparent commented:
@spackbot fix style

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent

bernhardkaindl reviewed (dismissed):
Agreed, `cmake` often picks up system packages which then may fail the `spack` build.

@becker33 @alalazo @haampie: Could you review and approve this separation?

johnwparent commented:
> Agreed, `cmake` often picks up system packages which then may fail the `spack` build.
> 
> @becker33 @alalazo @haampie: Could you review and approve this separation?

Thanks @bernhardkaindl 

FYI for anyone reviewing, this shouldn't be merged until the TODOs at the top are all resolved and the WIP designation is removed.
This _should_ be a draft PR but Gitlab won't run if it is, and I need to validate that I'm not breaking CMake based detection for the rest of Spack.

eugeneswalker commented:
Thank you for working on this!

johnwparent committed:
Finalize initial sandboxing

johnwparent commented:
@spackbot fix style

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent

johnwparent commented:
Adding the `X_DIR` and `X_ROOT` scrubbing seems to have caused GL failures where as the initial CMake sandboxing did not. 
Some quick grepping shows a large number of packages do rely on this behavior to allow consumers to find them, so that makes some sense, however given that the order of `setup_dependent_build_environment` invocation is leaf up, and that CMake is generally a leaf node in these cases, I wouldn't expect this to be the cause.

The intended behavior is to sanitize the environment from `X_DIR`, etc vars coming from the user env, while letting packages leverage the functionality of `X_DIR` in CMake.

The majority of the failure additionally seem to be include issues, rather than detection issues during configure as I would have expected, so this requires further investigation.

In the meantime, the initial basic sandboxing is good to go in, so if we want to piecemeal this effort, that's no issue.

johnwparent commented:
cc @scheibelp let me know what you think

scheibelp commented:
> Adding the X_DIR and X_ROOT scrubbing seems to have caused GL failures where as the initial CMake sandboxing did not.
> Some quick grepping shows a large number of packages do rely on this behavior to allow consumers to find them, so that makes some sense

Do you think it would be worth explicitly always setting `X_ROOT` (i.e. instead of unsetting it, we force those variables to point to the Spack-based installations)?

If you'd prefer to experiment with this (or some other approach to address these build issues) later, then I agree we can split up the sandboxing effort into multiple PRs.

johnwparent commented:
> > Adding the X_DIR and X_ROOT scrubbing seems to have caused GL failures where as the initial CMake sandboxing did not.
> > Some quick grepping shows a large number of packages do rely on this behavior to allow consumers to find them, so that makes some sense
> 
> Do you think it would be worth explicitly always setting `X_ROOT` (i.e. instead of unsetting it, we force those variables to point to the Spack-based installations)?

The thought had occurred to me, but I initially decided against it as we already set the `CMAKE_PREFIX_PATH` so setting these paths shouldn't change anything, unless a package needs extra logic, in which case that should be done by the individual packages.

That said, these failures are not what I'd expect, so I'll give that a try.

> If you'd prefer to experiment with this (or some other approach to address these build issues) later, then I agree we can split up the sandboxing effort into multiple PRs.

I think for a little experimentation I'm ok waiting to merge the parts of this PR that pass CI, but if it becomes pressing or this languishes I'll break things down a bit

johnwparent committed:
Try explicitly setting x_ROOT and x_DIR

johnwparent committed:
Repair relation between c-blosc and lz4

With CMake setup to prefer configs c-blosc now needs to
be able to use lz4's and lz4 needs to export a proper
export interface

This PR adds patches to the upstream commits that fix these issues.

johnwparent committed:
Update patch sha

johnwparent committed:
Update patch urls and shas

johnwparent commented:
This is now ready for general review (there will likely be a couple gitlab failures over the next few days but those will be resolved quickly/more easily than before)

haampie commented:
can you rebase then?

haampie commented:
Can you remove duplicate `CMAKE_FIND_USE_PACKAGE_REGISTRY`? It's done better above.

Also check if other defines only work from certain cmake versions on so that they don't end up in warnings.

haampie commented on lib/spack/spack/build_systems/cmake.py:
```diff
@@ -375,6 +382,39 @@ def std_args(pkg, generator=None):
+                    # by default these are secondary to the find modules
+                    # Ensure CMake prefers Spack derived modules always
+                    # by prefering config modules
+                    define("CMAKE_FIND_PACKAGE_PREFER_CONFIG", True),
```
This does not sound like a good idea to me... There are cases where FindZLIB.cmake should be used because Spack built zlib does not have an associated config file, whereas it may exist in a directory of an external. The status quo is no cmake config file for zlib, so the better default is FindZLIB.cmake.

The build system may also expect it to use FindZLIB.cmake which exports different variables than a ZLIB config file (iirc).

I don't see what this define has to do with sandboxing, so better to submit it as a separate PR if you want this change?

johnwparent replied:
I addressed this _exact_ concern in comments in multiple places in this PR. Please read them.

> There are cases where FindZLIB.cmake should be used because Spack built zlib does not have an associated config file, whereas it may exist in a directory of an external.

This is addressed inherently by the use of this CMake argument. The _`PREFER`_ section of the argument name correctly implies that this will only _prefer_ config modules, it will not use them exclusively. It simply inverts the order of the search to look for config modules first, and failing to find them, falls back on find modules rather than the opposite default behavior.

> so the better default is FindZLIB.cmake

The better default IMO is whatever relies exclusively on Spack derived contents, vendored find modules can do the wrong thing, this exact case w.r.t zlib is what motivated this PR. This PR does not preclude the use of the find modules, it prefers the configs.

> The build system may also expect it to use FindZLIB.cmake which exports different variables than a ZLIB config file (iirc).

This is a very valid concern, and one of which I thought of and addressed by means of a per pacakge opt out of this sandboxing. Initially I was adding patches/upstream submissions to fix these packages, but that grew infeasible pretty quickly. I realize the opt out is cludge-y, so if you have constructive thoughts, I'd be happy to hear them.

> define has to do with sandboxing

It's "sandboxing" in the sense that we're prefering configs (which are generally derived directly from Spack packages) and do not bring in anything but Spack based information over vendored find modules which do not originate from Spack based behavior and can accidentally bring in system info.

> so better to submit it as a separate PR if you want this change?

this feels arbitrary based on a very specific and narrow view of sandboxing

haampie replied:
I know what prefer means :) I also think I'm right in that this config option is orthogonal to sandboxing.

Consider this where it makes things only worse:

Try `fedora:40` with a Spack installed zlib mounted, run `dnf install zlib-devel cmake`, create the following `CMakeLists.txt` file:

```cmake
cmake_minimum_required(VERSION 3.20)
project(Example C)
find_package(ZLIB REQUIRED)
```

and run:

```
cmake . --debug-find -DCMAKE_PREFIX_PATH=/path/to/spack/.../zlib-ng-2.2.1-2uvj6hw3ioeq5x6phso6gcxgbqcqpomo -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON
```

it finds system zlib.

Run:

```
cmake . --debug-find -DCMAKE_PREFIX_PATH=/path/to/spack/.../zlib-ng-2.2.1-2uvj6hw3ioeq5x6phso6gcxgbqcqpomo
```

it finds Spack zlib.

Would you consider every package that depends on zlib to opt out of sandboxing then? :)

haampie commented:
I'm confused by this PR. The only contribution is adding `-DCMAKE_FIND_USE_SYSTEM_PATH:BOOL=FALSE`. The rest was already contributed, and not really related to build isolation.

johnwparent commented:
> I'm a bit confused by this PR. The contribution is basically adding `-DCMAKE_FIND_USE_SYSTEM_PATH:BOOL=FALSE`.

There are several other modifications being made in this PR...

> The bit about disabling `<pkg>_ROOT` is in principle redundant due to
> https://github.com/spack/spack/blob/8892c878ce173022c0efb18b35f1b21afd7a072b/lib/spack/spack/build_environment.py#L229-L234

I'd argue my implementation is both more robust and more performant, although yes, it is redundant, we should remove the code you linked.

> . The rest was already contributed

One line of several was already contributed...

> https://github.com/spack/spack/blob/4de5b664cddc0c923db27b2e2f43ab95938fc976/lib/spack/spack/build_systems/cmake.py#L101-L107

Ah I didn't notice that, that is better handling, will remove.

> , or not really related to build isolation.

I think we disagree on how this relates to build isolation.

> Also, what cannot be tested in Gitlab CI, and may be a blocker for this PR is that if you configure an external, and it is found through a cmake config file, then this config file itself may `find_package(...)` its transitive dependencies, which are necessarily system packages. Will that break?

The answer is "it depends, and I'm actively working on the handling of externals with relation to this PR". 

In some scenarios, i.e. Spack package A depends on B, which is external and depends on C which is a non Spack aware system library, this is fine, we're not sandboxing the PATH as of yet, so the library would likely be detected exactly the same way it was before unless it was relying on C_DIR being set in the environment or is an entry in B's `CMAKE_FIND_ROOT_PATH` variable, in which case, yes, that would be an issue. There also exists an issue with transitive external dependencies, which although not unique to this PR and at the moment an issue with Spack's handling of externals, would be exacerbated by this PR quite a bit. 

My current thought is to sandbox the PATH and carefully restore it being more considerate of Spack's transitive external dependencies. In that case, I would say the issue with the relation between packages A, B, and C, is one of modeling rather than sandboxing.

haampie commented:
> I'd argue my implementation is both more robust and more performant

In principle it's not terrible for a package to explicit do `setup_dependent_build_environment(...): env.set("<NAME>_ROOT=<some path>")` if the directory structure is non-standard. Idk if there are use cases, but it sounds like it could be useful if it forces cmake to search in a specific dir.
